src/base/bittorrent/session.h

```diff
@@ -92,7 +92,8 @@ namespace BitTorrent
         {
             Default = 0,
             MMap = 1,
-            Posix = 2
+            Posix = 2,
+            Pread = 3
```

Comment:
What is the different between these two? I'm under the impression that they are eventually the same.

Reply:
Not even close, they are completely different under the hood.
Posix has horrendous performance, uses synchronous I/O, don't support files preallocation etc.

Reply:
OK. Do you what functions do they call? Obviously one is pread/pwrite and the other?

Reply:
BTW it seems the old POSIX option has a bad name now because pread/pwrite is also defined in POSIX standard.

Reply:
> Do you know what functions do they call?

[`std::fread`](https://github.com/arvidn/libtorrent/blob/2ab8fedabb7716947edc77958da6d0b5e1040df9/src/posix_storage.cpp#L208) and [`std::fwrite`](https://github.com/arvidn/libtorrent/blob/2ab8fedabb7716947edc77958da6d0b5e1040df9/src/posix_storage.cpp#L271C36-L271C47).

> BTW it seems the old POSIX option has a bad name

That how libtorrent calls it. 🤷‍♀️

Reply:
> > Do you know what functions do they call?
> 
> [`std::fread`](https://github.com/arvidn/libtorrent/blob/2ab8fedabb7716947edc77958da6d0b5e1040df9/src/posix_storage.cpp#L208) and [`std::fwrite`](https://github.com/arvidn/libtorrent/blob/2ab8fedabb7716947edc77958da6d0b5e1040df9/src/posix_storage.cpp#L271C36-L271C47).

Thank you!

Reply:
> > BTW it seems the old POSIX option has a bad name
> 
> That's how libtorrent calls it. 🤷‍♀️

Maybe we could rename `Pread` option to `Mmap fallback`? The new name will not confuse between "POSIX-compliant" and pread. Also, the pread could be an implementation detail that we shouldn't advertise.

Reply:
> Maybe we could rename `Pread` option to `Mmap fallback`?

Maybe. I named it accounting future changes from arvidn/libtorrent#7013, as said above:

> This option also could be repurposed in the future, when the actual pread-disk-io would be in production.

Again, that's how libtorrent calls it. But I don't mind the naming here.

Reply:
> Maybe we could rename the `Pread` option to `Mmap fallback`?

IMO, `Mmap fallback` is completely meaningless/confusing to be used here. Its meaning is precisely that it does not use memory mapping at all, so why mention it?

Reply:
In my old commit made to test a new disk IO subsystem, I named the corresponding option more abstractly from the implementation details "Multi-threaded file read/write". I would use something like it.

Reply:
>so why mention it?

https://github.com/qbittorrent/qBittorrent/pull/21300#discussion_r1746670336
1. To differentiate it with "POSIX-compliant" option.
2. The pread could be an implementation detail that we shouldn't advertise.

>This option also could be repurposed in the future, when the actual pread-disk-io would be in production.

By only looking at its name, some users might be mistaken that the current option is the unfinished pread-disk-io.
And when we switched it to pread-disk-io under the hood, users might still think it is the old one.

Either we add more info/rename the "POSIX-compliant" one or use some obscure name like "Mmap fallback".

Reply:
>Either we add more info/rename the "POSIX-compliant" one or use some obscure name like "Mmap fallback".

Or both. Such as "POSIX-compliant (fread/fwrite)" and "Mmap fallback (pread/pwrite)".
As for pread-disk-io it would be just "Pread" as libtorrent calls it https://github.com/arvidn/libtorrent/pull/7013/files#diff-a5b3e85fb91ec27c5c24aff0a6756978bc0cbacaf78333c902ee63afb641b25dR886

Reply:
> By only looking at its name, some users might be mistaken that the current option is the unfinished pread-disk-io.
And when we switched it to pread-disk-io under the hood, some users might still think it is the old one.

So we talk only about UI option name or internal name too?

Reply:
> Such as "POSIX-compliant (fread/fwrite)" and "Mmap fallback (pread/pwrite)"

I don't think users are interested in underlying functions. And `POSIX-compliant` name is already widely used.

New option should not contain "mmap" word imo, as it is misleading. Better call it something like "Fallback IO" or something.

Reply:
>So we talk only about UI option name or internal name too?

Both. IMO they should be distinctively named.

Reply:
`DiskIOType::MMapFallback` works for internal name I think.

As for user-facing name... maybe we make it straight: `Disable memory mapped files`. It nicely summarizes the whole thing to the user.

Reply:
> New option should not contain "mmap" word imo, as it is misleading.

👍 
I strongly insist to not use "mmap" in its name.

Reply:
> "Multi-threaded file read/write"

This name would kinda imply that other options don't support multithreading. Which is not true of course.
And the amount of threads is controlled by respective option (`Asynchronous I/O threads`). If you set it to 1 it is not multithreaded anymore.

Reply:
> I strongly insist to not use "mmap" in its name.

What about negation? "Not mmap" aka "Disable memory mapped files"?

Reply:
> The pread could be an implementation detail that we shouldn't advertise.

In fact, it's hard to avoid mentioning some internal details in the name of these options, because they are essentially low-level settings. I would just not mention the unimportant details, like the fact that we use the same thing to provide different types of I/O. By enabling/disabling memory mapping, we essentially get two different I/O types (mmap based and pread/pwrite based).

> > This option also could be repurposed in the future, when the actual pread-disk-io would be in production.
>
>By only looking at its name, some users might be mistaken that the current option is the unfinished pread-disk-io.

We do not say that there is a one-to-one correspondence with some low-level subsystem of libtorrent. There is always an opportunity to be misleading when it comes to some advanced settings. In any case, we cannot provide all information through the option name. Details can be specified somewhere in documentation, in option tooltip, in the feature announcement on the official website, etc.

>And when we switched it to pread-disk-io under the hood, some users might still think it is the old one.

When we replace it with a new pread/pwrite based implementation, we can present it as an improvement of the old one.

Reply:
Soo, effectively you are saying to keep it as is? And what about the internal name?

Reply:
>In any case, we cannot provide all information through the option name. Details can be specified somewhere in documentation, in option tooltip, in the feature announcement on the official website, etc.

Then I would just call the new option "Customized" and explain the details in the tooltip.

>When we replace it with a new pread/pwrite based implementation, we can present it as an improvement of the old one.

The old one is not a proper disk io type. It is mmap type's fallback behavior. Present it as an improvement would be false advertising.

The whole naming issue arose is because it is modifying something instead of a proper disk io type itself. If obscure names such as "Customized" is still objected. Then the only way forward is to present the `mmap_file_size_cutoff` and `disk_write_mode` options instead of disguising them as a new disk io type.

Reply:
> "Customized"

I like that variant actually.

> Then the only way forward is to present the `mmap_file_size_cutoff` and `disk_write_mode` options instead of disguising them as a new disk io type.

I see a problem with that, as stated above. https://github.com/qbittorrent/qBittorrent/pull/21300#issuecomment-2332279378

Reply:
> The whole naming issue arose is because it is modifying something instead of a proper disk io type itself.

This is exactly "proper disk io type" from an external point of view. You just can't abstract away from the unimportant implementation details of to accept it. I can just copy its code, hardcodely disable memory mapping (and even delete all related code), what would you call it in that case? A "proper disk io type" or not?

Reply:
>This is exactly "proper disk io type" from an external point of view. You just can't abstract away from the unimportant implementation details of to accept it. I can just copy its code, hardcodely disable memory mapping (and even delete all related code), what would you call it in that case? A "proper disk io type" or not?

If you provided a `disk_io_constructor_type` then it is a genuine disk io type. For our case here it is just modified Mmap.

Reply:
Maybe call it that? `Modified memory mapped files (minimal use)`

Reply:
> `Modified memory mapped files (minimal use)`

The thing is, the mmap is disabled completely, there is no minimal use. That's why I and @glassez insist on not including mmap into the description.

On the other hand, I don't think users are very interested in internals. So we might as well simply lie about its true nature and call it something like `Memory mapped files (optimized)`.

Reply:
> > "Customized"
> 
> I like that variant actually.

I still think this is the best compromise here. Maybe even `Custom optimized`?
Very neutral wording, then we could explain what exactly it does in the docs.

Reply:
`optimized` doesn't sound very accurate as it's not clear which way the optimization is going, `minimal use` is closer to that. After all, these are really parameters for power users.

Reply:
> The thing is, the mmap is disabled completely, there is no minimal use. That's why I and @glassez insist on not including mmap into the description.

Memory mapping based thing that doesn't use memory mapping completely is nonsense.

Reply:
> `optimized` doesn't sound very accurate as it's not clear which way the optimization is going,

Yeah, that's a valid point.

> `minimal use` is closer to that.

But in that case, I don't see how `minimal use` is better than plain simple `disabled` though.

> After all, these are really parameters for power users.

Also true. I personally don't mind how it's called.
But the more concise name we pick, the easier it will be to talk with people about it.

Reply:
>Maybe even Custom optimized?
>
>minimal use is closer to that. After all, these are really parameters for power users.

Or `Custom simplified`? as it is reduced from mmap + fallback to only fallback.

Reply:
> Memory mapping based thing that doesn't use memory mapping completely is nonsense.

And if we still want to expose technical details, I would say `Force pread/pwrite` sounds good to me.

Reply:
`Modified memory mapped files (disabled)` as an option, yes, but it sounds really weird.

Reply:
>I would say Force pread/pwrite sounds good to me.

I could settle on this. At least it is implying it has tweaked behavior.

Reply:
Ok then, `DiskIOType::MMapFallback` + `Force pread/pwrite`.

Reply:
If you want to refer to these two IO types as the same, but with enabled/disabled memory mapping, the main part of name should not contain "memory mapping", for example:
1. Regular (memory mapping enabled)
2. Regular (memory mapping disabled)

("Regular" is just an example.)

Reply:
I guess that's outside of the scope of this PR.

Reply:
> > I would say Force pread/pwrite sounds good to me.
> 
> I could settle on this. At least it is implying it has tweaked behavior.

I really don't understand that. One option uses memory mapping, another uses pread/pwrite. Why mention the negligent details?

As about `DiskIOType::MMapFallback`.
What life cycle do you think for this option? Should it be replaced by a new one when it becomes available? Or you want to leave both?
In any case, this is a user-visible string used in the configuration file so I wouldn't use `mmap` in it as well.

Reply:
> In any case, this is a user-visible string used in the configuration file so I wouldn't use `mmap` in it as well.

What do you suggest then? Something closer to the user-visible string?

Reply:
>Or you want to leave both?

Yes, until it is known that the new one don't have nasty bugs or the old one is inferior than the new one in every way.

Reply:
> > In any case, this is a user-visible string used in the configuration file so I wouldn't use `mmap` in it as well.
> 
> What do you suggest then? Something closer to the user-visible string?

I just want the names of the options to reflect their distinctive features (one type of I/O uses memory mapped files, the other uses regular I/O feature), and not unimportant details (like the fact that both are provided by the same libtorrent classes). The mention of memory mapping in the name of the discussed I/O type (both UI and internal), which does not involve the use of such nemory mapping, is unacceptable.

Reply:
Understood. I'll make them in sync then. It's not like we have technical restrictions about it.